[doc] Logstash Kubernetes - Persistent Storage Docs (#14714)
elastic/logstash: docsk8s/setting-up/ls-k8s-persistent-storage.asciidoc

[[ls-k8s-persistent-storage]]
=== Stateful {ls}

WARNING: This documentation is still in development and may be changed or removed in a future release.

When running {ls} with certain use cases, {ls} needs to persist data to disk: 

- Enable persistent queue to absorb bursts of events 
- Enable dead letter queue for the corrupted events that cannot be processed.

> - Some plugins save and resume the state of work

Author: This section wants to say there are three situations logstash need to persist data to disk, so the user needs to use `StatefulSet`. 
1. PQ
2. DLQ
3. Particular plugins such as jdbc-input, file-input

Keep them in list form is appropriate

          volumeMounts:
            - name: logstash-data <2>
              mountPath: /usr/share/logstash/data
            - name: logstash-pipeline
              mountPath: /usr/share/logstash/pipeline
            - name: logstash-config
              mountPath: /usr/share/logstash/config/logstash.yml
              subPath: logstash.yml
            - name: logstash-config
              mountPath: /usr/share/logstash/config/pipelines.yml
              subPath: pipelines.yml
      volumes:
        - name: logstash-pipeline
          configMap:
            name: logstash-pipeline
        - name: logstash-config
          configMap:
            name: logstash-config
  volumeClaimTemplates: <1>
    - metadata:
        name: logstash-data
        labels:
          app: logstash-demo
      spec:
        accessModes: ["ReadWriteOnce"]
        resources:
          requests:
            storage: 2Gi
--

Everything is similar to `Deployment`, except the usage of `VolumeClaimTemplates`.

<1> Request 2G of persistent storage from `PersistentVolumes`.

<2> Mount the storage to `/usr/share/logstash/data`. This is the default path of {ls} and its plugins for any persistence needs.

> NOTE: The feature of persistent volume expansion depends on the storage class, checkout your cloud provider.

Author: Yes, I can give a [link](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/)

[[ls-k8s-persistent-storage]]
=== Stateful {ls}

WARNING: This documentation is still in development and may be changed or removed in a future release.

When you are running {ls} for certain use cases, you need {ls} to persist data to disk.

- Enable persistent queue (PQ) to absorb bursts of events 
- Enable dead letter queue (DLQ) to accept corrupted events that cannot be processed.

> - Some plugins save and resume the state of work

Author: tested locally. the format looks good

[[ls-k8s-persistent-storage]]
=== Stateful {ls}

WARNING: This documentation is still in development and may be changed or removed in a future release.

> In some scenarios, {ls} need to persist data to disk: 

Reviewer: ```suggestion
When running {ls} with certain use cases, {ls} needs to persist data to disk: 
```

- Enable persistent queue to absorb bursts of events 
- Enable dead letter queue for the corrupted events that cannot be processed.
- Some plugins save and resume the state of work

> Kubernetes scheduler can shutdown pods at anytime and spawn the process to another node. To against data loss during unexpected termination, use `StatefulSet` to deploy {ls}.

Reviewer: What do you think about giving additional context, and stating that we should use `StatefulSet` instead of `Deployment`

```suggestion
In these cases, we need to ensure that we can preserve state, and take account that the Kubernetes scheduler can shutdown pods at anytime and spawn the process to another node. To enable this, we define our {ls} deployment using `StatefulSet` rather than `Deployment`.
```

Reviewer: By additional context, I meant making it clear that `StatefulSet` was a replacement for `Deployment` and indicating the changes that need to be made, and your changes do this

      spec:
        accessModes: ["ReadWriteOnce"]
        resources:
          requests:
            storage: 2Gi
--

Everything is similar to `Deployment`, except the usage of `VolumeClaimTemplates`.

<1> Request 2G of persistent storage from `PersistentVolumes`.

<2> Mount the storage to `/usr/share/logstash/data`. This is the default path of {ls} and its plugins for any persistence needs.

NOTE: The feature of persistent volume expansion depends on the storage class, checkout your cloud provider.

[[persistent-storage-pq]]
==== Persistent queue (PQ)

To enable {logstash-ref}/persistent-queues.html[PQ] for every pipeline, specify options in `logstash.yml`. 

[source,yaml]
--
apiVersion: v1
kind: ConfigMap
metadata:
  name: logstash-config
data:
  logstash.yml: |
    api.http.host: "0.0.0.0"
    queue.type: persisted
    queue.max_bytes: 1024mb
...
--

To specify options per pipeline, set in `pipelines.yml`.

[source,yaml]
--
apiVersion: v1
kind: ConfigMap
metadata:
  name: logstash-config
data:
  logstash.yml: |
    api.http.host: "0.0.0.0"
  pipelines.yml: |
    - pipeline.id: fast_ingestion
      path.config: "/usr/share/logstash/pipeline/fast.conf"

>       queue.type: persisted

Author: I would avoid adding complexity to the setup. Making logstash runnable on k8s may be challenging enough for some users. I try to make the necessary steps as few as possible for easy digestion.

Given the way of deployment in container and the pipeline definition are from `ConfigMap`, I see less practice in changing `paht.config`

      queue.max_bytes: 1024mb
    - pipeline.id: slow_ingestion
      path.config: "/usr/share/logstash/pipeline/slow.conf"
      queue.type: persisted
      queue.max_bytes: 2048mb
--

> Queue data store in `/usr/share/logstash/data/queue` by default.

Author: I moved it to the PQ description, believe users understand the path is the default and has no difference to global or per-pipeline.

> [[persistent-storage-pq-util]]

Author: 💡 good idea!

===== Run PQ utility

In case the pod is terminated ungracefully and hence the persistent queues are corrupted, `pqcheck` and `pqrepair` are the tools for troubleshooting.

To run {logstash-ref}/persistent-queues.html#pqcheck[pqcheck] to identify corrupted files: 

[source,bash]
--
kubectl exec logstash-0 -it -- /usr/share/logstash/bin/pqcheck /usr/share/logstash/data/queue/pipeline_id
--

To run {logstash-ref}/persistent-queues.html#pqrepair[pqrepair] to repair the queue: 

[source,bash]
--
kubectl exec logstash-0 -it -- /usr/share/logstash/bin/pqrepair /usr/share/logstash/data/queue/pipeline_id
--

[[persistent-storage-pq-drain]]
===== Draining the queue

Set `queue.drain: true` to drain and wait for the queues to empty when receive SIGTERM. However, the default `TerminationGracePeriodSeconds` in Kubernetes is 30 seconds. If {ls} is still running after the grace period elapsed, Kubernetes send SIGKILL to terminated the pod immediately. Hence, {ls} cannot finish the draining.

A workaround is to set a very long period like 1 year to `TerminationGracePeriodSeconds` to make sure {ls} get enough time to drain all events.

[[persistent-storage-dlq]]
==== Dead letter queue (DLQ)

> To enable {logstash-ref}/dead-letter-queues.html[dead letter queue], specify options in `logstash.yml`. The default path of DLQ is `/usr/share/logstash/data/dead_letter_queue`.

Reviewer: Do we need to mention that the queue contents themselves will be in `/usr/share/logstash/data/dead_letter_queue/<pipeline_id>`

Author: probably no. When we config a different path for DLQ, we also did not include <pipeline_id>.

[source,yaml]
--
apiVersion: v1
kind: ConfigMap
metadata:
  name: logstash-config
data:
  logstash.yml: |
    api.http.host: "0.0.0.0"
    dead_letter_queue.enable: true <1>
  pipelines.yml: |
    - pipeline.id: main <2>
      path.config: "/usr/share/logstash/pipeline/main.conf"
    - pipeline.id: dlq <3>
      path.config: "/usr/share/logstash/pipeline/dlq.conf"
--

<1> Enable DLQ for all pipelines that use {logstash-ref}/plugins-outputs-elasticsearch.html[elasticsearch output plugin]

> <2> The `main` pipeline generates fail events to DLQ. Checkout the pipeline definition in the next session.

Reviewer: ```suggestion
<2> The `main` pipeline sends failed events to DLQ. Checkout the pipeline definition in the next section.
```

> <3> The `dlq` pipeline consumes the events in DLQ, fixes the error and sends the events to {es} again. Checkout the pipeline definition in the next session.

Reviewer: ```suggestion
<3> The `dlq` pipeline should consume events from the DLQ, fix errors and re-send events to {es}. Checkout the pipeline definition in the next section.
```

[source,yaml]
--
apiVersion: v1
kind: ConfigMap
metadata:
  name: logstash-pipeline
data:
  main.conf: | <1>
    input {
      exec {
        command => "uptime"
        interval => 5
      }
    }
    output {
      elasticsearch { 
        hosts => ["https://hostname.cloud.es.io:9200"]
        index => "uptime-%{+YYYY.MM.dd}"
        user => 'elastic'
        password => 'changeme'
      }
    }
  dlq.conf: | <2>
    input {
      dead_letter_queue {
        path => "/usr/share/logstash/data/dead_letter_queue"
        commit_offsets => true
        pipeline_id => "main"
      }
    }
    filter {
        # Do your fix here
    }
    output {
      elasticsearch { 
        hosts => ["https://hostname.cloud.es.io:9200"]
        index => "dlq-%{+YYYY.MM.dd}"
        user => 'elastic'
        password => 'changeme'
      }
    }
--

> <1> An example pipeline that insert value to {es}. In the test environment, to generate fail event manually, use {ref}/indices-close.html[_close] API to close the index.

Reviewer: ```suggestion
<1> An example pipeline that tries to send events to a closed index in {es}. To test this functionality manually, use {ref}/indices-close.html[_close] API to close the index.
```

> <2> The pipeline use {logstash-ref}/plugins-inputs-dead_letter_queue.html[dead_letter_queue input plugin] to consume DLQ events. Add filter plugins to fix error causing fail insertion, for example, mapping error.

Reviewer: ```suggestion
<2> This pipeline use {logstash-ref}/plugins-inputs-dead_letter_queue.html[dead_letter_queue input plugin] to consume DLQ events. This example sends to a different index, but you can add filter plugins to fix other types of error causing fail insertion, such as mapping errors.
```

[[persistent-storage-plugins]]
==== Plugins that require local storage to track work done

> 

Reviewer: Maybe an introduction here:

```suggestion
Many Logstash plugins are stateful, and need to use persistent storage to track the current state of the work that they are doing. 

```

> Here is the list of plugins require persistent storage to track the state of work. In order to have them work properly, use `StatefulSet` with `VolumeClaimTemplates`, checkout <<persistent-storage-statefulset>>.

Reviewer: I don't think this is exhaustive, so maybe easier to give a rule of how to identify, and a list of "popular" plugins

```suggestion
Logstash plugins that are stateful will typically have some kind of `path` that needs to be configured, such as `sincedb_path` or `last_run_metadata_path`

Here is the list of popular plugins that will require persistent storage, and the use of a `StatefulSet` with `VolumeClaimTemplates`, checkout <<persistent-storage-statefulset>>.
```
